src/main/java/engineering/swat/watch/impl/JDKPoller.java

```diff
@@ -0,0 +1,120 @@
+     * Linux can be thrown into a deadlock if you try to start 1000 threads and then do a register at the same time.
+     */
+    private static final ExecutorService registerPool = Executors.newFixedThreadPool(nCores);
+
+    static {
+        try {
+            service = FileSystems.getDefault().newWatchService();
+        } catch (IOException e) {
+            throw new RuntimeException("Could not start watcher", e);
+        }
+        // kick off the poll loop
+        poll();
+    }
+
+    private static void poll() {
+        try {
+            WatchKey hit;
+            while ((hit = service.poll()) != null) {
+                logger.trace("Got hit: {}", hit);
+                try {
+                    var watchHandler = watchers.get(hit);
+                    if (watchHandler != null) {
+                        var events = hit.pollEvents();
+                        logger.trace("Found watcher for hit: {}, sending: {} (size: {})", watchHandler, events, events.size());
+                        watchHandler.accept(events);
+                    }
+                }
+                catch (Throwable t) {
+                    logger.catching(Level.INFO, t);
+                    // one exception shouldn't stop all the processing
+                }
+                finally{
+                    hit.reset();
+                }
+            }
+
+        }
+        finally {
+            // schedule next run
+            // note we don't want to have multiple polls running in parallel
+            // so that is why we only schedule the next one after we're done
+            // processing all messages
+            CompletableFuture
+                .delayedExecutor(1, TimeUnit.MILLISECONDS)
+                .execute(JDKPoller::poll);
+        }
+    }
+
+
+    public static Closeable register(Path path, Consumer<List<WatchEvent<?>>> changes) throws IOException {
+        logger.debug("Register watch for: {}", path);
+
+        try {
+            return CompletableFuture.supplyAsync(() -> {
+                    try {
+                        return path.register(service, ENTRY_CREATE, ENTRY_MODIFY, ENTRY_MODIFY, ENTRY_DELETE);
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    }
+                }, registerPool) // read registerPool why we have to add a limiter here
+                .thenApplyAsync(key -> {
+                    watchers.put(key, changes);
+                    return new Closeable() {
+                        @Override
+                        public void close() throws IOException {
+                            logger.debug("Closing watch for: {}", path);
+                            key.cancel();
+                            watchers.remove(key);
+                        }
+                    };
+                })
+                .get(); // we have to do a get here, to make sure the `register` function blocks
```

Comment:
it's not meant to be async, it's just meant to throttle stuff so that all the concurrent calls to `JDKPoller::register` don't hammer the `Path::register` function to a stall. On linux this happens 50% of the time. So we add a small statically sized thread-pool to make sure we don't have too many open `Path::register` calls.